internal/route/api/v1/repo/tree.go

```diff
@@ -0,0 +1,65 @@
+package repo
+
+import (
+	"fmt"
+
+	"github.com/gogs/git-module"
+
+	"gogs.io/gogs/internal/context"
+	"gogs.io/gogs/internal/db"
+)
+
+func GetRepoTree(c *context.APIContext) {
+	repoPath := db.RepoPath(c.Params(":username"), c.Params(":reponame"))
+	gitTree, err := c.Repo.GitRepo.GetTree(c.Params(":sha"))
+	if err != nil {
+		c.NotFoundOrServerError("GetRepoTree", git.IsErrNotExist, err)
+		return
+	}
+	entries, err := gitTree.ListEntries()
+	if err != nil {
+		c.ServerError("GetRepoTree", err)
+		return
+	}
+	var results []map[string]interface{}
```

Comment:
Use `[]map[string]interface{}` in this case is not ideal, should define concrete types:

```go
type repoGitTreeResponse struct {
	Sha  string `json:"sha"`
	URL  string `json:"url"`
	Tree []*repoGitTree `json:"tree"`
}

type repoGitTree struct{
	Path string `json:"path"`
	Mode string `json:"mode"`
	Type string `json:"type"`
	Size int    `json:"size,omitempty"`
	Sha  string `json:"sha"`
	URL  string `json:"url"`
}
```

_Generated using https://mholt.github.io/json-to-go/_

Reply:
I agree, these should be now implemented, verified that these still return the same response:
```bash
{
  "sha": "c59441ded1549b149def0d4c54594d31a7f3718f",
  "url": "http://localhost:3000/api/v1/repos/root/testrepo/git/trees/c59441ded1549b149def0d4c54594d31a7f3718f",
  "tree": [
    {
      "path": "/home/bharatnc/gogs-repositories/root/testrepo.git",
      "mode": "120000",
      "type": "blob",
      "size": 1077,
      "sha": "472ac2361b65136b393d652de25341e2ea44f299",
      "url": "http://localhost:3000/api/v1/repos/root/testrepo/git/trees/472ac2361b65136b393d652de25341e2ea44f299"
    },
    {
      "path": "/home/bharatnc/gogs-repositories/root/testrepo.git",
      "mode": "120000",
      "type": "blob",
      "size": 12,
      "sha": "70fcb456d436f08462602f26df6fb7e167e7a916",
      "url": "http://localhost:3000/api/v1/repos/root/testrepo/git/trees/70fcb456d436f08462602f26df6fb7e167e7a916"
    },
    {
      "path": "/home/bharatnc/gogs-repositories/root/testrepo.git",
      "mode": "120000",
      "type": "blob",
      "size": 380,
      "sha": "092c58d4b63df5779a4d020b1fdbb762421bbb4f",
      "url": "http://localhost:3000/api/v1/repos/root/testrepo/git/trees/092c58d4b63df5779a4d020b1fdbb762421bbb4f"
    }
  ]
}

```